src/test/java/ch/uzh/ifi/hase/soprafs23/controller/UserControllerTest.java

```diff
@@ -43,61 +48,136 @@ public class UserControllerTest {
-        .andExpect(jsonPath("$.name", is(user.getName())))
         .andExpect(jsonPath("$.username", is(user.getUsername())))
         .andExpect(jsonPath("$.status", is(user.getStatus().toString())));
+
+    // verify that the correct calls on UserService were made
+    verify(userService, times(1)).createUser(Mockito.any(User.class));
   }
 
+
+  @Test
+  public void postUsers_returnsErrorConflict_whenInvalidInput() throws Exception {
+    // create UserPostDTO for the request body
+    UserPostDTO userPostDTO = new UserPostDTO();
+    userPostDTO.setUsername(user.getUsername());
+    userPostDTO.setPassword(user.getPassword());
+
+    // mocks the createUser(user) method in UserService
+    given(userService.createUser(Mockito.any(User.class)))
+        .willThrow(new ResponseStatusException(HttpStatus.CONFLICT, "error message"));
+
+    // when
+    MockHttpServletRequestBuilder postRequest = post("/users")
+                                                .contentType(MediaType.APPLICATION_JSON)
+                                                .content(asJsonString(userPostDTO));
+
+    // then
+    mockMvc.perform(postRequest)
+        .andExpect(status().isConflict());
+
+    // verify that the correct calls on UserService were made
+    verify(userService, times(1)).createUser(Mockito.any(User.class));
+  }
+
+    /**
+     * TEST PUT user by username - VALID version
+     */
+    @Test
+    public void loginUser_valid() throws Exception {
+        // create new User
+        User user = new User();
+        user.setId(3L);
+        user.setUsername("testUsername");
+
+        //don't need comparison with internal representation here
+
+        // when/then -> do the request - just trying to put (=update) sth.
+        MockHttpServletRequestBuilder putRequest = put("/users/testUsername/login")
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(asJsonString(user));
+
+        // then
+        mockMvc.perform(putRequest)
+                .andExpect(status().isOk());
+    }
+
+    /**
+     * TEST PUT user by username - INVALID version, username not found
+     */
+    @Test
+    public void loginUser_INVALID() throws Exception {
+        User user = new User();
+        user.setId(4L);
+        user.setUsername("test");
+
+
+        //try to get user with ID 4
+        // We are mocking the method get userbyID here
+        given(userService.getUserByUsername("testUsername")).willThrow(new ResponseStatusException(HttpStatus.NOT_FOUND, "There is no user with this username."));
+
+        // when/then -> do the request
+        MockHttpServletRequestBuilder putRequest = put("/users/testUsername/login")
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(asJsonString(user));
+
+        // then
+        mockMvc.perform(putRequest)
+                .andExpect(status().isNotFound()); // has to map again with above http status
+    }
+
```

Comment:
another error it can throw is the 400 for incorrect password -> should get an own test as well